numba/utils.py

```diff
@@ -291,6 +290,16 @@ def shutting_down(globals=globals):
     return v is True or v is None
 
 
+# weakref.finalize registers an exit function that runs all finalizers for
+# which atexit is True. Some of these finalizers may call shutting_down() to
+# check whether the interpreter is shutting down. For this to behave correctly,
+# we need to make sure that _at_shutdown is called before the finalizer exit
+# function. Since atexit operates as a LIFO stack, we first contruct a dummy
+# finalizer then register atexit to ensure this ordering.
+weakref.finalize(lambda: None, lambda: None)
+atexit.register(_at_shutdown)
```

Comment:
I looked in the CPython source... I think:
1. ` weakref.finalize(lambda: None, lambda: None)` will create and initialise a new finalizer object, in the process of initialisation it will register its own classmethod for invoking finalizers with `atexit`. If a `finalize` object has already been created elsewhere this mechanism will already have run and an internal flag in the `weakref.finalize` impl will prevent re-running. Therefore either the above or some other call to `weakref.finalize` will ensure the classmethod for invoking finalizers is registered with `atexit` by this point in the import sequence.
2. `atexit.register(_at_shutdown)` will register the `_at_shutdown` function with `atexit`, as the `atexit` stack is LIFO, this will be executed ahead of the `weakref.finalize` class method registered in 1. This means that `_at_shutdown` will always have run ahead of any finalizer that may need information about this state via a call to `shutting_down()`.

Basically, this seems correct!

Reply:
Thanks - that mirrors my thinking (and also I reached the same conclusion via the same method)!